### PR TITLE
kvevent: implement chunked blocking buffer

### PR DESCRIPTION
**Pull request by jayshrivastava**
### **kvevent: refactor blocking buffer benchmark**
This change updates the blocking buffer micro benchmark
in several ways:
- it uses different types of events
- it uses more producers than consumers to keep the
  buffer full
- it makes b.N correspond to the total number of events,
  so the benchmark can analyze allocs per event

Release note: None

Release justification: This change updates a benchmark only.

### **kvevent: implement chunked buffer event queue**
This change implements a simple chunked event queue.
The purpose of this queue is to be used by
kvevent.blockingBuffer in subsequent commits.

Release note: None

Release justification: This change does not affect
any production code. It adds files which are not
called by any packages.

### **kvevent: refactor memory buffer to chunked linked list**
This change refactors kvevent/blocking_buffer.go to use
a chunked linked list instead of a regular linked list to
reduce pointer usage. Note that the underlying sync.Pool,
which is also a linked list, will use less pointers due to
us pooling chunks instead of events.

Release note: None

Release justification: This change significantly
improves performance by significantly reducing
pressure on GC. Consequently, this significantly
improves foreground SQL p99 latency. GC has
been causing severe issues in production changefeeds.
Merging this change in this release is worth it
for its potential to reduce incidents.

### **Results (micro)**
These are the results of running the microbenchmark.
`./dev bench pkg/ccl/changefeedccl/kvevent --filter=BenchmarkMemBuffer --count=10 --bench-mem  --stream-output  --test-args="--test.benchtime=45s" -- --nocache_test_results --test_verbose_timeout_warnings |& tee bench.txt`
```
name          old time/op    new time/op    delta
MemBuffer-10    1.22µs ± 2%    0.85µs ± 3%  -30.04%  (p=0.000 n=8+10)

name          old alloc/op   new alloc/op   delta
MemBuffer-10     0.00B          0.00B          ~     (all equal)

name          old allocs/op  new allocs/op  delta
MemBuffer-10      0.00           0.00          ~     (all equal)
```
- Memory usage is 0 due to pooling in both implementations. 
- We can achieve a higher throughput with the chunked implementation - about 50-60M events in 45 seconds as opposed to ~40M with the old implementation. 

### **Results (Macro)**
Full results are published [here](https://docs.google.com/document/d/10-Pc_cnkCUhkRFpLyBAId39q1aqsjuqTb2k4ehTmQCY/edit?usp=sharing). In summary:

I analyzed performing by running TPC-C for 30 mins on a 15 node cluster with 10k warehouses. Before starting the workload, I started a changefeed on the `order_line` table (~200GB). I also set the following cluster settings to stress the buffer and pressure GC:
`changefeed.backfill.concurrent_scan_requests = 100;`
`changefeed.memory.per_changefeed_limit = '1073741824';` (~1GB)

Then, I analyzed SQL latency from admin UI and GC performance using the output of `GODEBUG=gctrace=1.` These are the outcomes:
- The p99 SQL latency during the workload was reduced from approx. 1.75s -> 0.150s (91%)
- CPU time spent doing GC was reduced from 37.86 mins -> 20.75 mins (45%)
- The p99 spike at the beginning of the workload was reduced from approx. 15s -> 12s (20%)

### Relevant Issues
Addresses: https://github.com/cockroachdb/cockroach/issues/84582
(for now...)

### CODE REVIEW

File: pkg/ccl/changefeedccl/kvevent/blocking_buffer.go [hunk 1]
```diff
@@ -24,6 +23,13 @@ import (
 	"github.com/cockroachdb/errors"
 )
 
+type queue interface {
+	enqueue(e Event)
+	dequeue() (Event, bool)
+	empty() bool
+	purge()
+}
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/kvevent/blocking_buffer.go:31]: Please drop this interface and use chunk queue directly; I only added this interface
when testing alternative implementations.

File: pkg/ccl/changefeedccl/kvevent/alloc.go [hunk 2]
```diff
@@ -84,6 +94,15 @@ func (a *Alloc) Merge(other *Alloc) {
...
 func (a *Alloc) clear()       { *a = Alloc{} }
 func (a *Alloc) isZero() bool { return a.ap == nil }
+func (a *Alloc) init(bytes int64, p pool) {
+	if !a.isZero() {
+		panic("cannot initialize already initialized alloc")
+	}
+	a.bytes = bytes
+	a.entries = 1
+	a.ap = p
+
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/kvevent/alloc.go:104]: nit: drop extra \n

File: pkg/ccl/changefeedccl/kvevent/blocking_buffer.go [hunk 3]
```diff
@@ -68,22 +79,24 @@ func NewMemBuffer(
 		AbstractPool: quotapool.New("changefeed", quota, opts...),
 		metrics:      metrics,
 	}
+	b.mu.queue = &bufferEventChunkQueue{}
+	// b.mu.queue = &eventRing{}
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/kvevent/blocking_buffer.go:83]: stray comment line.

File: pkg/ccl/changefeedccl/kvevent/event.go [hunk 4]
```diff
@@ -100,40 +100,40 @@ type Event struct {
 }
 
 // Type returns the event's Type.
-func (b *Event) Type() Type {
-	if b.kv.Key != nil {
+func (e *Event) Type() Type {
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/kvevent/event.go:103]: I would revert this change and do this minor refactoring in separate PR.
(not that I mind it, it's just such refactors make subsequent backports more painful).

**Review by miretskiy — APPROVED** [review]
Mostly nits; giving LGTM, but please remove unneeded interface, and revert unnecessary refactors from this PR.

**Comment by miretskiy** [comment]
@nvanbenschoten : FYI: thanks to your investigations, @jayshrivastava made these changes that show significant improvements.

**Comment by nvanbenschoten** [comment]
This is great! Nice experimentation @jayshrivastava.

> Fixes: https://github.com/cockroachdb/cockroach/issues/84709

I don't think I see where we're addressing #84709 in this PR. Doesn't the `Event` struct still contain 10 pointers? So doesn't a `bufferEventChunk` contain `128x10+1=1281` pointers?

**Comment by miretskiy** [comment]
> I don't think I see where we're addressing #84709 in this PR. Doesn't the `Event` struct still contain 10 pointers? So doesn't a `bufferEventChunk` contain `128x10+1=1281` pointers?

I think previous changes  (https://github.com/cockroachdb/cockroach/commit/e8e664cd16f8d488c9ca7036c50ae9c7035e966d#diff-a2e21a39cea12e1823c4c6f7ce7e1513214a575ff049f204cf5633284cf8c6c9) replaced resolved events which were allocating a pointer.  
Event does contains various slices (key/value/resolved span) but those should be coming from rangefeed; so I don't think we are allocating them.

**Comment by nvanbenschoten** [comment]
I think there might be a bit of confusion here. The concern isn't reducing heap allocations, it's about reducing the cost of GC by eliminating pointers that need to be traversed during the GC mark-sweep phase. So https://github.com/cockroachdb/cockroach/commit/e8e664cd16f8d488c9ca7036c50ae9c7035e966d#diff-a2e21a39cea12e1823c4c6f7ce7e1513214a575ff049f204cf5633284cf8c6c9 might have actually hurt, as the inlining replaced 1 pointer (`*ResolvedSpan`) for 2 (`ResolvedSpan.Span.Key` and `ResolvedSpan.Span.EndKey`).

**Comment by miretskiy** [comment]
> replaced 1 pointer (`*ResolvedSpan`) for 2 (`ResolvedSpan.Span.Key` and `ResolvedSpan.Span.EndKey`).

This is what's confusing to me too: because those 2 pointers needed to be deleted anyway because those were allocated
by rangefeed RPC. So.. I don't know if we're making situation worse.  I think this is an improvement because instead of 3 pointers that need to be traversed, it's 2 (which were allocated anyway).

**Comment by nvanbenschoten** [comment]
Does every `Event` have a non-empty resolved span? My claim that it might make things worse is based on the assumption that the vast majority of `Event`s aren't carrying a resolved span, but if that's not true then I agree that https://github.com/cockroachdb/cockroach/commit/e8e664cd16f8d488c9ca7036c50ae9c7035e966d#diff-a2e21a39cea12e1823c4c6f7ce7e1513214a575ff049f204cf5633284cf8c6c9 should help.

**Comment by miretskiy** [comment]
No, not every event has non-empty resolved span; Can't really make a claim that majority of events aren't carrying resolved
span (you could have pretty low traffic table where majority of events will be resolved span events).
I don't know if just having 16 bytes in the structure (to hold resolve spans) is necessarily bad -- those arrays are allocated, contiguously.  

We could try to basically break up event further: 
for "regular" event, we need 3 []byte slices (key, value, prev value).  For Resolved, we only need 2.
So, have just 2 slices, plus type of the event (we already have "flush" boolean -- which can be replace
with some sort of uint8).  I don't know if it's worth the complexity. But we should cha.

**Comment by shermanCRL** [comment]
Looks like we need a manual backport? Let’s get that going if we haven’t already.

**Comment by jayshrivastava** [comment]
@miretskiy and I are looking to test out the changes against TPC-E before backporting. It would be nice to see the impact of these changes to the same workload Nathan ran originally.

**Comment by jayshrivastava** [comment]
@shermanCRL Just finished testing with TPC-E, with and without this change. Please see the appendix in [this doc](https://docs.google.com/document/d/10-Pc_cnkCUhkRFpLyBAId39q1aqsjuqTb2k4ehTmQCY/edit#) more info. 

Summary (with both a **massive** TPC-E load and a **massive** changefeed backfill running):
- SQL latency is reduced from 560ms -> 400ms (29%)
- SQL statement throughput goes up from 5k -> 6k (20%)

@miretskiy and I have more ideas regarding https://github.com/cockroachdb/cockroach/issues/84709. I think it makes the most sense to carry them out and backport everything at the same time.

**Comment by shermanCRL** [comment]
Nice! What’s the baseline SQL latency _without_ changefeeds?

**Comment by jayshrivastava** [comment]
40ms with only TPC-E running

**Pull request merged**